Added ghidra for android reversing (#1500)
OWASP/owasp-mstg: Document/0x05c-Reverse-Engineering-and-Tampering.md

Dalvik and ART both support the Java Native Interface (JNI), which defines a way for Java code to interact with native code written in C/C++. As on other Linux-based operating systems, native code is packaged (compiled) into ELF dynamic libraries (\*.so), which the Android app loads at run time via the `System.load` method. However, instead of relying on widely used C libraries (such as glibc), Android binaries are built against a custom libc named [Bionic](https://github.com/android/platform_bionic "Bionic libc"). Bionic adds support for important Android-specific services such as system properties and logging, and it is not fully POSIX-compatible.

> For reversing an Android applicaiton containing native code, we need to understand couple of data structures related to JNI bridge between Java and native code. From reversing perspective, we need to be aware of two key data structures, `JavaVM` and `JNIEnv`. Both of these data structures are pointers to pointers to function tables. `JavaVM` provides an interface to invoke functions to create and destroy a JavaVM. Android allows only one `JavaVM` per process, thus this data structure is not relevant for us. While the other data structure `JNIEnv` provides access to most the JNI functions (by pointing to pointers of these functions). These functions are accessible at a fixed offset through the `JNIEnv` pointer. This `JNIEnv` pointer is the first parameter passed to every JNI function. This concept is discussed again with an example in subsequent chapters.

Reviewer: ```suggestion
When reversing an Android application containing native code, we need to understand a couple of data structures related to the JNI bridge between Java and native code. From the reversing perspective, we need to be aware of two key data structures: `JavaVM` and `JNIEnv`. Both of them are pointers to pointers to function tables.
- `JavaVM` provides an interface to invoke functions for creating and destroying a JavaVM. Android allows only one `JavaVM` per process and is not really relevant for our reversing purposes.
- `JNIEnv` provides access to most of the JNI functions which are accessible at a fixed offset through the `JNIEnv` pointer. This `JNIEnv` pointer is the first parameter passed to every JNI function. We will discuss this concept again with the help of an example later in this chapter.
```

> It is worth highlighting that for reversing the native code in an Android applicaiton we will need a disassembler. And later in the chapter while analyzing the application we will observe that analyzing disassembled native code is more challenging than disassembled Java code.

Reviewer: ```suggestion
It is worth highlighting that analyzing disassembled native code is much more challenging than disassembled Java code. When reversing the native code in an Android application we will need a disassembler.
```

JNIEXPORT jstring JNICALL Java_sg_vantagepoint_helloworld_MainActivity_stringFromJNI(JNIEnv *env, jobject)
```

> So where is the native implementation of this function? If you look into the `lib` directory of the APK archive, you'll see eight subdirectories named after different processor architectures. Each of these directories contains a version of the native library `libnative-lib.so` that has been compiled for the processor architecture in question. When `System.loadLibrary` is called, the loader selects the correct version based on the device that the app is running on. Before moving ahead, take a special note of the first parameter passed to the current JNI function. It is the same `JNIEnv` data structure which was discussed earlier in the chapter.

Reviewer: ```suggestion
So where is the native implementation of this function? If you look into the "lib" directory of the unzipped APK archive, you'll see several subdirectories (one per supported processor architecture), each of them containing a version of the native library, in this case `libnative-lib.so`. When `System.loadLibrary` is called, the loader selects the correct version based on the device that the app is running on. Before moving ahead, pay attention to the first parameter passed to the current JNI function. It is the same `JNIEnv` data structure which was discussed earlier in this section.
```

###### radare2

> To open the file in radare2 you only have to run `r2 -A HelloWord-JNI/lib/armeabi-v7a/libnative-lib.so`. The chapter "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md/#android-basic-security-testing "Android Basic Security Testing")" already introduced radare2. Remember that you can use the flag `-A` to run the `aaa` command right after loading the binary in order to _analyze all referenced code_.

Reviewer: Anchor is not required to refer to the chapter itself :)

```suggestion
To open the file in radare2 you only have to run `r2 -A HelloWord-JNI/lib/armeabi-v7a/libnative-lib.so`. The chapter "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md "Android Basic Security Testing")" already introduced radare2. Remember that you can use the flag `-A` to run the `aaa` command right after loading the binary in order to _analyze all referenced code_.
```

╰           0x00000e82      1047           bx r2
```

Notice that in this case we're not starting with the `-A` flag not running `aaa`. Instead, we just tell radare2 to analyze that one function by using the _analyze function_ `af` command. This is one of those cases where we can speed up our workflow because you're focusing on some specific part of an app.

> The workflow can further be improved by using [r2ghidra-dec](https://github.com/radareorg/r2ghidra-dec "r2ghidra-dec"), a deep Ghidra decompiler integration for radare2, and generate corresponding decompiled C code. The decompiled C code can aid in quick analysis of the binary.    

Reviewer: Very nice remark ;)

```suggestion
The workflow can be further improved by using [r2ghidra-dec](https://github.com/radareorg/r2ghidra-dec "r2ghidra-dec"), a deep integration of Ghidra decompiler for radare2. r2ghidra-dec generates decompiled C code, which can aid in quickly analyzing the binary.    
```

When this function returns, R0 contains a pointer to the newly constructed UTF string. This is the final return value, so R0 is left unchanged and the function returns.

###### Ghidra

> On opening the library in Ghidra we can see all the function define in the **Symbol Tree** in **Functions** tree. The native library for the current application is relatively very small. There are three user defined functions - `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper functioning of the shared library. The instructions in the function `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI` are already discussed in detail in previous sections. In this section we can look into the decompilation of the function. 

Reviewer: ```suggestion
After opening the library in Ghidra we can see all the functions defined in the **Symbol Tree** panel under **Functions**. The native library for the current application is relatively very small. There are three user defined functions: `FUN_001004d0`, `FUN_0010051c`, and `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI`. The other symbols are not user defined and are generated for proper functioning of the shared library. The instructions in the function `Java_sg_vantagepoint_helloworldjni_MainActivity_stringFromJNI` are already discussed in detail in previous sections. In this section we can look into the decompilation of the function. 
```

> Inside the current function there is a call to another function, whose address is obtained by accessing an offset in JNIEnv pointer. This logic has been diagrammatically demonstrated above as well. Corresponding C code for the disassembled function is shown in the decompilation window. This decompiled C code makes it much easier to understand the function call made. Since this function is small and not complex, the decompilation output is accurate, this can change drastically when dealing with complex functions.

Reviewer: ```suggestion
Inside the current function there is a call to another function, whose address is obtained by accessing an offset in the `JNIEnv` pointer (found as `plParm1`). This logic has been diagrammatically demonstrated above as well. The corresponding C code for the disassembled function is shown in the **Decompiler** window. This decompiled C code makes it much easier to understand the function call being made. Since this function is small and extremely simple, the decompilation output is very accurate, this can change drastically when dealing with complex functions.
```

Reviewer: Is the `JNIEnv` pointer explained before? If not could you make the user understand this? This is a very interesting point to highlight when reversing Android native libs. I've just searched the chapter for JNIEnv and it's mentioned but not explained. Maybe you could take the chance to explain it in the first location where it's being referred (around line 180?). That would be very helpful so that the reader can understand what's behind it and why is this offset `0x538` being added to it ;)

Author: Sure, will fix it.

##### Debugging with jdb

> The `adb` command line tool was introduced in the "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md/#android-basic-security-testing "android Basic Security Testing")" chapter. You can use its `adb jdwp` command to list the process ids of all debuggable processes running on the connected device (i.e., processes hosting a JDWP transport). With the `adb forward` command, you can open a listening socket on your host machine and forward this socket's incoming TCP connections to the JDWP transport of a chosen process.

Reviewer: ```suggestion
The `adb` command line tool was introduced in the "[Android Basic Security Testing](0x05b-Basic-Security_Testing.md "Android Basic Security Testing")" chapter. You can use its `adb jdwp` command to list the process IDs of all debuggable processes running on the connected device (i.e., processes hosting a JDWP transport). With the `adb forward` command, you can open a listening socket on your host machine and forward this socket's incoming TCP connections to the JDWP transport of a chosen process.
```
